### PR TITLE
feat: implement new experimental login workflow for local development

### PR DESCRIPTION
**Pull request by emmenko**
**Disclaimer: this is an opt-in experimental feature. Use it at your own risk.**

> We want to test this feature internally first. Until then, we discourage you to try it out.

TL;DR: this feature changes a bit how developers log in when developing the application. It also requires to provide additional information in the custom application config.
These changes are part of a bigger topic related to new changes and improvements around Custom Applications. So I'm not going to spoil too much.

Stay tuned for more upcoming information.

### CODE REVIEW

File: packages/application-config/schema.json [hunk 1]
```diff
@@ -17,10 +17,6 @@
     }
   },
   "properties": {
-    "id": {
```
Comment (emmenko) [inline packages/application-config/schema.json:20]: This has moved to the `env.production.applicationId` field (it wasn't used before anyway).

File: packages/application-config/src/process-config.ts [hunk 2]
```diff
@@ -49,15 +47,25 @@ const processConfig = ({
   const additionalAppEnv = appConfig.additionalEnv ?? {};
   const revision = (additionalAppEnv.revision as string) ?? '';
 
-  const appEnvConfig = isProd ? appConfig.env.production : developmentConfig;
+  // Feature flags
+  const isOidcForDevelopmentEnabled =
+    processEnv.ENABLE_OIDC_FOR_DEVELOPMENT === 'true' ||
+    // @ts-expect-error
+    processEnv.ENABLE_OIDC_FOR_DEVELOPMENT === true;
```
Comment (emmenko) [inline packages/application-config/src/process-config.ts:54]: Opt-in feature

File: packages/application-shell/src/apollo-links/header-link.ts [hunk 3]
```diff
@@ -68,10 +68,16 @@ const headerLink = new ApolloLink((operation, forward) => {
...
+    STORAGE_KEYS.SESSION_TOKEN
+  );
 
   operation.setContext({
     credentials: 'include',
-    headers: omitEmpty({
+    headers: omitEmpty<Headers>({
+      [SUPPORTED_HEADERS.AUTHORIZATION]: sessionToken
+        ? `Bearer ${sessionToken}`
+        : undefined,
```
Comment (emmenko) [inline packages/application-shell/src/apollo-links/header-link.ts:80]: Here we send the session token as  an Authorization header.

File: packages/application-shell/src/components/application-shell/redirect-to-login.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    `${STORAGE_KEYS.NONCE}_${nonce}`,
+    JSON.stringify(state)
+  );
+  return nonce;
+};
+
+const RedirectToLogin = () => {
+  const location = useLocation();
+
+  if (window.app.__DEVELOPMENT__) {
```
Comment (emmenko) [inline packages/application-shell/src/components/application-shell/redirect-to-login.tsx:36]: For local development, we send the construct the authorize request and redirect to the real login page.

File: packages/application-shell/src/components/authenticated/auth-callback.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        console.warn(
+          `Cannot parse session state for "${key}".\n${unparsedSessionState}`
+        );
+      }
+    }
+  }
+  return null;
+};
+
+const AuthCallback = () => {
```
Comment (emmenko) [inline packages/application-shell/src/components/authenticated/auth-callback.tsx:48]: This is the component that handles the auth callback redirect (after login). Most of the things here are very similar to what we also do for the SSO callback.

File: cypress/integration/playground/application-shell.js [hunk 6]
```diff
@@ -1,51 +1,75 @@
...
-import { URL_BASE, URL_STATE_MACHINES } from '../../support/urls';
+import {
+  URL_BASE,
+  URL_STATE_MACHINES,
+  ENTRY_POINT_STATE_MACHINES,
+} from '../../support/urls';
 
 describe('when user is authenticated', () => {
+  beforeEach(() => {
+    cy.loginByOidc({ entryPointUriPath: ENTRY_POINT_STATE_MACHINES });
```
Comment (emmenko) [inline cypress/integration/playground/application-shell.js:11]: This is a new command (see `@commercetools-frontend/cypress` package)

File: cypress/integration/playground/application-shell.js [hunk 7]
```diff
@@ -1,51 +1,75 @@
...
+  });
   it('should log out with reason "user"', () => {
-    cy.login({ redirectToUri: URL_STATE_MACHINES });
-
     cy.findByRole('button', { name: /open user settings menu/i }).click();
-    cy.findByRole('link', { name: /logout/i }).click();
```
Comment (emmenko) [inline cypress/integration/playground/application-shell.js:11]: We don't have a logout page anymore, so we simply check that the link is correct (as it redirects to another domain)

File: packages/application-shell/src/components/application-shell-provider/application-shell-provider.tsx [hunk 8]
```diff
@@ -56,11 +57,12 @@ const ApplicationShellProvider = <AdditionalEnvironmentProperties extends {}>(
               <Router history={ApplicationShellProvider.history}>
                 <GtmBooter trackingEventList={props.trackingEventList || {}}>
                   <Authenticated
+                    locale={browserLocale}
+                    applicationMessages={props.applicationMessages}
```
Comment (emmenko) [inline packages/application-shell/src/components/application-shell-provider/application-shell-provider.tsx:61]: Needed for rendering the "error page" for the `/auth/callback` route.

File: packages/application-shell/src/components/authenticated/auth-callback-error-page.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+  box-sizing: border-box;
+  width: 100%;
+  margin: 0;
+  border: 0;
+  border-style: solid;
+  border-top-width: 1px;
+  border-top-color: ${customProperties.colorNeutral};
+`;
+
+const AuthCallbackErrorPage = (props: TProps) => {
```
Comment (emmenko) [inline packages/application-shell/src/components/authenticated/auth-callback-error-page.tsx:42]: The structure is very similar to our internal "PublicPageContainer" component.

File: packages/application-shell/src/components/authenticated/auth-callback-error-page.tsx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                <Card>
+                  <Spacings.Stack scale="l">
+                    <Spacings.Inline justifyContent="center">
+                      <img
+                        src={FailedAuthenticationSVG}
+                        alt="Failed authentication"
+                      />
+                    </Spacings.Inline>
+                    <Text.Headline as="h2">
+                      {'Authentication error'}
```
Comment (emmenko) [inline packages/application-shell/src/components/authenticated/auth-callback-error-page.tsx:66]: I don't think we need to translate these messages for now:

* it's still an experimental feature
* it's only used for development.

File: packages/application-shell/src/components/authenticated/auth-callback.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        console.warn(
+          `Cannot parse session state for "${key}".\n${unparsedSessionState}`
+        );
+      }
+    }
+  }
+  return null;
+};
+
+const AuthCallback = (props: TProps) => {
```
Comment (emmenko) [inline packages/application-shell/src/components/authenticated/auth-callback.tsx:50]: If you're familiar with the MC SSO callback component, this is very similar: we parse the token from the hash fragment, decode it, validate the nonce, etc.

File: packages/application-shell/src/components/authenticated/has-cached-authentication-state.ts [hunk 12]
```diff
@@ -1,6 +1,53 @@
+import type { ApplicationWindow } from '@commercetools-frontend/constants';
+
 import { STORAGE_KEYS } from '../../constants';
+import { buildOidcScope } from './helpers';
+
+declare let window: ApplicationWindow;
 
-const hasCachedAuthenticationState = () =>
-  window.localStorage.getItem(STORAGE_KEYS.IS_AUTHENTICATED) === 'true';
+const hasCachedAuthenticationState = (): boolean => {
```
Comment (emmenko) [inline packages/application-shell/src/components/authenticated/has-cached-authentication-state.ts:8]: I should probably add some unit tests for this function.

File: packages/application-shell/src/components/project-switcher/project-switcher.tsx [hunk 13]
```diff
@@ -168,11 +172,19 @@ const ProjectSwitcher = (props: Props) => {
         aria-labelledby="project-switcher"
         onChange={(event) => {
           const selectedProjectKey = event.target.value;
-          if (selectedProjectKey !== props.projectKey)
+          if (selectedProjectKey !== props.projectKey) {
+            if (window.app.__DEVELOPMENT__) {
+              window.localStorage.setItem(
+                STORAGE_KEYS.ACTIVE_PROJECT_KEY,
+                selectedProjectKey
+              );
```
Comment (emmenko) [inline packages/application-shell/src/components/project-switcher/project-switcher.tsx:180]: Doing this triggers a new login flow because the session scope has changed.

File: packages/application-shell/src/components/redirector/redirector.tsx [hunk 14]
```diff
@@ -9,29 +9,41 @@ import { location } from '../../utils/location';
...
+  queryParams?: QueryParams;
 };
 
-const redirectTo = (targetUrl: string) => location.replace(targetUrl);
+const redirectTo = (targetUrl: string) => {
+  location.replace(targetUrl);
+};
 
 const Redirector = (props: Props) => {
-  React.useEffect(() => {
```
Comment (emmenko) [inline packages/application-shell/src/components/redirector/redirector.tsx:22]: We don't need all those variables to be defined inside the `useEffect`, hence this change.

File: packages/application-shell/src/configure-store.ts [hunk 15]
```diff
@@ -51,10 +54,23 @@ const mergeObjectValues = <T>(object: { [key: string]: T }) =>
...
+    [SUPPORTED_HEADERS.X_APPLICATION_ID]: window.app.applicationId,
+    [SUPPORTED_HEADERS.X_TEAM_ID]: selectTeamIdFromLocalStorage(),
+  });
+};
+
 const sdkMiddleware = createSdkMiddleware({
   getCorrelationId: patchedGetCorrelationId,
   getProjectKey: selectProjectKeyFromUrl,
-  getTeamId: selectTeamIdFromLocalStorage,
+  getAdditionalHeaders,
```
Comment (emmenko) [inline packages/application-shell/src/configure-store.ts:73]: We're passing now a different function as it better encapsulates the required headers.

File: packages/cypress/src/add-commands/index.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+  'loginByOidc',
+  ({ entryPointUriPath }: { entryPointUriPath: string }) => {
+    Cypress.log({ name: 'loginByOidc' });
+
+    const projectKey = Cypress.env('PROJECT_KEY');
+    const sessionNonce = uuidv4();
+
+    cy.task('customApplicationConfig', {
+      entryPointUriPath,
+    }).then((appConfig: ApplicationConfig['env']) => {
```
Comment (emmenko) [inline packages/cypress/src/add-commands/index.ts:22]: We're using a [task](https://docs.cypress.io/api/commands/task.html) to be able to perform nodejs operations, specifically using the filesystem.

Essentially we want to read up the runtime environment config (which is what is available at `window.app`) as the first thing, as we need to get some of the values to construct the request to fetch a token.

File: packages/cypress/src/add-commands/index.ts [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+              `${STORAGE_KEYS.NONCE}_${sessionNonce}`,
+              JSON.stringify({ applicationId, query: {} })
+            );
+            win.sessionStorage.setItem(
+              STORAGE_KEYS.SESSION_SCOPE,
+              sessionScope
+            );
+          },
+        });
+      });
```
Comment (emmenko) [inline packages/cypress/src/add-commands/index.ts:62]: The token response contains a redirect link. We're visiting that link, which contains the session token as hash fragments, etc.

Before doing that though, we need to prepare the browser storage with some values, as otherwise the auth callback will fail to verify that the request is legit.

File: playground/package.json [hunk 18]
```diff
@@ -5,9 +5,9 @@
   "private": true,
   "scripts": {
     "build": "mc-scripts build",
-    "start": "dotenv -- mc-scripts start",
+    "start": "dotenv -c development -- mc-scripts start",
```
Comment (emmenko) [inline playground/package.json:8]: https://www.npmjs.com/package/dotenv-cli#cascading-env-variables

File: packages/application-shell/src/components/authenticated/has-cached-authentication-state.ts [hunk 19]
```diff
@@ -1,6 +1,53 @@
...
+        return false;
+      }
+      const activeProjectKey = window.localStorage.getItem(
+        STORAGE_KEYS.ACTIVE_PROJECT_KEY
+      );
+      if (activeProjectKey) {
+        // The application is not requesting a project key, therefore
+        // we assume that the application does not need a project context
+        // and we can remove the cached project key.
+        // This is usually the case for applications like account.
```
Comment (adnasa) [inline packages/application-shell/src/components/authenticated/has-cached-authentication-state.ts:25]:
```suggestion
        // GIVEN The application is not requesting a project key,
        // THEN we assume that the application does not need a project context
        // GIVEN The application is not requesting a project key,
        // we remove the cached project key.
        // This is the case of an application like `account`.
```

File: packages/application-shell/src/components/authenticated/helpers.ts [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+  if (options.projectKey) {
+    projectClaims.push(`${OIDC_CLAIMS.PROJECT_KEY}:${options.projectKey}`);
+    projectClaims.push(
+      ...(window.app.__DEVELOPMENT__?.permissions?.view ?? []).map(
+        (scope) => `${OIDC_CLAIMS.VIEW}:${scope}`
+      )
+    );
+    projectClaims.push(
+      ...(window.app.__DEVELOPMENT__?.permissions?.manage ?? []).map(
+        (scope) => `${OIDC_CLAIMS.MANAGE}:${scope}`
```
Comment (adnasa) [inline packages/application-shell/src/components/authenticated/helpers.ts:20]: I expected this to be `<project-key>:<oauth-scope>` not `<scope>:<oauth-scope>`.
it looks like this is by design so I could've mixed things up

**Review by adnasa — COMMENTED** [review]
it's hard to follow the user flow on how works.
I doubt I can find time to test this out locally.

A walk through would be awesome.
thanks!

File: packages/application-config/src/process-config.ts [hunk 21]
```diff
@@ -49,15 +49,27 @@ const processConfig = ({
...
```
Comment (tdeekens) [inline packages/application-config/src/process-config.ts:56]: You might be able to just `JSON.parse(processEnv.ENABLE...)` to get the `string | boolean` into what you want: boolean.

File: packages/application-config/src/process-config.ts [hunk 22]
```diff
@@ -69,18 +81,45 @@ const processConfig = ({
...
       ...omitEmpty(additionalAppEnv),
-      ...(applicationId ? { applicationId } : {}),
+      // TODO: how else should we provide the app identifier?
+      applicationId,
       applicationName: appConfig.name,
       entryPointUriPath: appConfig.entryPointUriPath,
+      ...(!isProd && isOidcForDevelopmentEnabled
+        ? {
+            __DEVELOPMENT__: omitEmpty({
+              authorizeUrl: `${mcApiUrl.protocol}//${mcApiUrl.host.replace(
```
Comment (tdeekens) [inline packages/application-config/src/process-config.ts:113]: Should we explicitly define `authorizeUrl` or is there a benefit from inferring it.

File: packages/application-config/src/schema.ts [hunk 23]
```diff
@@ -26,14 +22,44 @@ export interface JSONSchemaForCustomApplicationConfigurationFiles {
    * The URL of the Merchant Center API. We usually recommend to use the `cloudIdentifier` option to avoid possible typos.
    */
   mcApiUrl?: string;
+  /**
+   * The permissions configuration for the required OAuth Scopes for this Custom Application.
+   */
+  permissions?: {
```
Comment (tdeekens) [inline packages/application-config/src/schema.ts:28]: Should we call this scopes? Just thinking out loud. Permissions infer scopes in the MC. This doesn't infer anything it's just scopes. Without the indirection of permissions.

File: packages/application-shell/src/apollo-links/header-link.ts [hunk 24]
```diff
@@ -21,7 +21,7 @@ declare let window: ApplicationWindow;
 type ApolloContextWithInMemoryCache = TApolloContext & {
   cache: ApolloClient<NormalizedCacheObject>;
 };
-
+type Headers = { [key: string]: string };
```
Comment (tdeekens) [inline packages/application-shell/src/apollo-links/header-link.ts:24]: Nit: `Record<string, string>`.

File: packages/application-shell/src/components/application-shell/redirect-to-login.tsx [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+
+    // According to the OIDC spec, the `state` parameter is recommended to be sent
+    // to the authorization server to help mitigating Cross-Site Request Forgery.
+    // https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
+    // Using `state` or `nonce` is very similar but with subtle differences.
+    // Here we follow the approach described by Auth0 on how to use both, where
+    // we generate a `nonce`, store it in session storage together with some state,
+    // and send it to the authorization server as the `state` parameter.
+    // https://auth0.com/docs/protocols/oauth2/redirect-users
+    // Additionally, we still send the `nonce` parameter as well.
```
Comment (tdeekens) [inline packages/application-shell/src/components/application-shell/redirect-to-login.tsx:52]: 👍🏻

File: packages/application-shell/src/components/application-shell/redirect-to-login.tsx [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+
+    // Store session scopes, to be able to detect if requested scopes changed
+    // in the application config and invalidate the session.
+    // This is only valid for local development.
+    window.sessionStorage.setItem(STORAGE_KEYS.SESSION_SCOPE, requestedScope);
+
+    return (
+      <Redirector
+        to="login/authorize"
+        origin={window.app.__DEVELOPMENT__.authorizeUrl}
```
Comment (tdeekens) [inline packages/application-shell/src/components/application-shell/redirect-to-login.tsx:71]: Can you elaborate why we scope this under `__DEVELOPERMENT__` and not e.g. `oidcConfig`?

File: packages/application-shell/src/components/application-shell/redirect-to-logout.tsx [hunk 27]
```diff
@@ -1,22 +1,42 @@
...
   const servedByProxy = useApplicationContext(
     (context) => context.environment.servedByProxy
   );
+
+  if (window.app.__DEVELOPMENT__) {
+    // Remove the `sessionToken` from storage, so that the AppShell can initiate
+    // a new authorization flow.
+    window.sessionStorage.removeItem(STORAGE_KEYS.SESSION_TOKEN);
+    window.sessionStorage.removeItem(STORAGE_KEYS.SESSION_SCOPE);
+    window.sessionStorage.removeItem(STORAGE_KEYS.IS_AUTHENTICATED);
```
Comment (tdeekens) [inline packages/application-shell/src/components/application-shell/redirect-to-logout.tsx:30]: Nit: Maybe we can have a small `development-oidc-localstorage.ts` module which allows setting and unsetting things. Just to not spread knowledge o the keys into different parts of the shell.

File: packages/application-shell/src/components/authenticated/authenticated.tsx [hunk 28]
```diff
@@ -1,23 +1,61 @@
...
+
+  // Fall back to check for cookie-based authentication.
   return <AmILoggedIn {...props} />;
 };
 Authenticated.displayName = 'Authenticated';
 
-export default Authenticated;
+const AuthenticationRoutes = (props: TProps) => (
+  <Switch>
+    <Route path={`/account/auth/callback`}>
```
Comment (tdeekens) [inline packages/application-shell/src/components/authenticated/authenticated.tsx:42]: `auth` or `authorize`? Is it the same or is `auth` preferred?

File: packages/application-shell/src/components/authenticated/helpers.ts [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+  if (teamId) {
+    projectClaims.push(`${OIDC_CLAIMS.TEAM_ID}:${teamId}`);
+  }
+  return [
+    // This is required as per OIDC spec.
+    OIDC_CLAIMS.OPEN_ID,
+    // Custom claims
+    ...projectClaims,
+  ].join(' ');
+};
```
Comment (tdeekens) [inline packages/application-shell/src/components/authenticated/helpers.ts:34]: How would we test store scopes with this? Is it intended or a possible scenario?

File: packages/cypress/README.md [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+## Usage
+
+This package extends Cypress' `cy` command.
+
+Add this line to your project's `cypress/plugins/index.js`:
+
+```javascript
+const customApplications = require('@commercetools-frontend/cypress/task');
+
+module.exports = (on, config) => {
```
Comment (tdeekens) [inline packages/cypress/README.md:24]: Not sure it applies but I still would recommend to install like this:

```js
customApplicationsPlugin.install(on, config);
```

This gives more freedom in the future. You can pass config options potentially and have access to the whole "on".

**Review by tdeekens — COMMENTED** [review]
Impressive! I did a first pass review but plan to do a second.

File: cypress/integration/playground/state-machines.js [hunk 31]
```diff
@@ -1,18 +1,20 @@
-import { URL_STATE_MACHINES, URL_STATE_MACHINES_ID } from '../../support/urls';
+import {
+  URL_STATE_MACHINES_ID,
+  ENTRY_POINT_STATE_MACHINES,
+} from '../../support/urls';
 
 describe('State machines', () => {
+  beforeEach(() => {
+    cy.loginByOidc({ entryPointUriPath: ENTRY_POINT_STATE_MACHINES });
```
[the remark this answers is not in the document]
Comment (emmenko) [reply cypress/integration/playground/state-machines.js:8]: The problem is when having multiple applications in the same repository.

File: packages/application-config/src/process-config.ts (hunk 22 above)
Comment (emmenko) [reply packages/application-config/src/process-config.ts:113]: We should infer it, because the environment (staging, production, etc) has to be the same with the APIs that the application uses.

File: packages/application-config/src/schema.ts (hunk 23 above)
Comment (emmenko) [reply packages/application-config/src/schema.ts:28]: Initially I went with this because it's more or less what we wanted to show in the UI (registration form).
We then changed it and went with view/manage oauth scopes.

I'm fine remaining this, if you think it's confusing. Thanks for the reminder.

File: packages/application-shell/src/components/application-shell/redirect-to-login.tsx (hunk 26 above)
Comment (emmenko) [reply packages/application-shell/src/components/application-shell/redirect-to-login.tsx:71]: Because it's a feature that should only be used in development (for now). So having `__DEVELOPMENT__` makes things more clear (that was the intention anyway).

Do you see other reasons for not using `__DEVELOPMENT__`?

File: packages/application-shell/src/components/application-shell/redirect-to-logout.tsx (hunk 27 above)
Comment (emmenko) [reply packages/application-shell/src/components/application-shell/redirect-to-logout.tsx:30]: Thanks, I'll see what I can do.

File: packages/application-config/src/schema.ts (hunk 23 above)
Comment (tdeekens) [reply packages/application-config/src/schema.ts:28]: I think it would help. The confusing bit could be otherwise that you might think that by adding "orders" here under "view" that I'd get the scopes the View Orders permission implies: e.g. types. That's not true tho.

File: packages/application-shell/src/components/application-shell/redirect-to-login.tsx (hunk 26 above)
Comment (tdeekens) [reply packages/application-shell/src/components/application-shell/redirect-to-login.tsx:71]: Not strictly. I just stumbled across it. Can't come up with a better indication of that either.

File: packages/application-shell/src/components/authenticated/authenticated.tsx (hunk 28 above)
Comment (emmenko) [reply packages/application-shell/src/components/authenticated/authenticated.tsx:42]: Good question.

Usually `/authorize` is well defined in the spec as the endpoint to start the authentication flow.
The `/auth/callback` is usually left to the user to decide.

I don't have a particular opinion about the name, it's also something we can manage internally.

Do you have suggestions? We can also change it to `/oidc/callback`

File: packages/application-shell/src/components/authenticated/helpers.ts (hunk 29 above)
Comment (emmenko) [reply packages/application-shell/src/components/authenticated/helpers.ts:34]: I think it can be defined as a normal scope.

Comment (tdeekens) [reply packages/application-shell/src/components/authenticated/helpers.ts:34]: I scope is `manage_orders:{projectKey}:{storeKey}`. So if I add `view.[0] = 'orders:storeA` would that work?

File: cypress/integration/playground/application-shell.js [hunk 32]
```diff
@@ -1,51 +1,75 @@
...
+
+describe('failed OIDC authentication', () => {
+  describe('when sessionToken is missing', () => {
+    it('should show auth callback error page', () => {
+      cy.visit(`/${URL_STATE_MACHINES}/auth/callback`);
+      cy.findByText('Authentication error');
+      cy.findByText(/missing sessionToken/i);
+      cy.percySnapshot();
+    });
+  });
```
[the remark this answers is not in the document]
Comment (emmenko) [reply cypress/integration/playground/application-shell.js:66]: Oh...well just check Percy 😅

File: packages/application-config/src/process-config.ts [hunk 33]
```diff
@@ -69,18 +81,45 @@ const processConfig = ({
...
+
+  // The real application ID is only used in production.
+  // In development, we prefix the entry point with the "__local" prefix.
+  // This is important to determine to which URL the MC should redirect to
+  // after successful login.
+  let applicationId: string | undefined = isOidcForDevelopmentEnabled
+    ? `__local:${appConfig.entryPointUriPath}`
+    : undefined;
+  if (isProd) {
+    // TODO: decide if we do require the application ID or not.
```
[the remark this answers is not in the document]
Comment (emmenko) [reply packages/application-config/src/process-config.ts:93]: Yes that was the idea. I still think it's valid, we can still discuss it maybe just to be sure.

File: packages/application-config/test/process-config.spec.js [hunk 34]
```diff
@@ -61,6 +62,7 @@ describe('processing a simple config', () => {
           applicationName: 'avengers-app',
           entryPointUriPath: 'avengers',
           cdnUrl: 'https://avengers.app/',
+          entryPointUriPath: 'avengers',
```
[the remark this answers is not in the document]
Comment (emmenko) [reply packages/application-config/test/process-config.spec.js:65]: Ah no, good catch

File: packages/application-shell/src/components/authenticated/authenticated.tsx (hunk 28 above)
Comment (tdeekens) [reply packages/application-shell/src/components/authenticated/authenticated.tsx:42]: I like `oidc/calback` to be honest. It clearly shows what auth flow the callback is for. Also if that deviates a bit from the norm.

File: packages/application-shell/src/components/authenticated/helpers.ts (hunk 20 above)
Comment (emmenko) [reply packages/application-shell/src/components/authenticated/helpers.ts:20]: Yes the idea is to have well-defined claims:
* `project_key`
* `view`
* `manage`

Having `<project-key>:<oauth-scope>` makes it harder to parse, because you always assume that the left-hand value is the project key.

File: packages/application-shell/src/components/authenticated/helpers.ts (hunk 29 above)
Comment (emmenko) [reply packages/application-shell/src/components/authenticated/helpers.ts:34]: I double checked. It's not supported at the moment.

I think we can fix this in the backend.

So you would still pass the oauth scopes as `view_products` or `view_orders:storeA`.

File: packages/cypress/README.md (hunk 30 above)
Comment (emmenko) [reply packages/cypress/README.md:24]: I double checked this. We're exporting a `task`, which already implies that in the `cypress/plugins/index.js` you have to use `on('task', {...})`.

It does not make much sense to pass `on` to the task itself.

If we were to implement a plugin, then I guess it makes sense to do as you suggested.

File: packages/application-config/src/schema.ts (hunk 23 above)
Comment (emmenko) [reply packages/application-config/src/schema.ts]: Changed it: https://github.com/commercetools/merchant-center-application-kit/pull/1934/commits/43934feb54e6fd7d2ae7ddbbf5cb60be557a5aec

File: packages/application-shell/src/components/authenticated/authenticated.tsx (hunk 28 above)
Comment (emmenko) [reply packages/application-shell/src/components/authenticated/authenticated.tsx]: https://github.com/commercetools/merchant-center-application-kit/pull/1934/commits/496d5add7b13982ba5e2847555bf08e95e75db18

File: packages/application-shell/src/components/application-shell/redirect-to-logout.tsx (hunk 27 above)
Comment (emmenko) [reply packages/application-shell/src/components/application-shell/redirect-to-logout.tsx]: https://github.com/commercetools/merchant-center-application-kit/pull/1934/commits/05a7dd9668a0285e0a4fef80e637c6be1c4cf0c3

**Review by tdeekens — APPROVED** [review]
💯 🏩 ⭐
